File: src/api/zig/proof.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// proof.zig — Proof Envelope Generation
//
// Every API response (GraphQL, gRPC, REST) is wrapped in a ProofEnvelope
// that provides tamper-evident hashing. Phase 1 uses "light" mode (SHA-256
// hash only). Phase 2+ will add Ed448 signatures ("full" mode).
//
// Replaces: proof.v

const std  = @import("std");
const t    = @import("types.zig");

/// Generate a UUID v4 string into `out` (must be at least 37 bytes).
/// Uses std.crypto.random for cryptographically secure bytes.
/// Sets the version 4 and variant bits per RFC 4122.
pub fn generateUuidV4(out: *[37]u8) void {
    var bytes: [16]u8 = undefined;
    std.crypto.random.bytes(&bytes);
    // Version 4: bits 12-15 of time_hi_and_version
    bytes[6] = (bytes[6] & 0x0f) | 0x40;
    // Variant bits: bits 6-7 of clk_seq_hi_res
    bytes[8] = (bytes[8] & 0x3f) | 0x80;

    _ = std.fmt.bufPrint(
        out,
        "{x:0>2}{x:0>2}{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}{x:0>2}{x:0>2}{x:0>2}{x:0>2}",
        .{
            bytes[0],  bytes[1],  bytes[2],  bytes[3],
            bytes[4],  bytes[5],
            bytes[6],  bytes[7],
            bytes[8],  bytes[9],
            bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15],
        },
    ) catch {
        // bufPrint cannot fail for a 37-byte buffer with exactly 36 chars
        // of formatted output; branch unreachable in practice.
        @memcpy(out[0..36], "00000000-0000-4000-8000-000000000000");
        out[36] = 0;
        return;
    };
    out[36] = 0;
}

/// Format the current UTC time as RFC 3339 into `out` (at least 32 bytes).
pub fn formatRfc3339(out: *[32]u8) void {
    const ts = std.time.timestamp();
    const epoch = std.time.epoch.EpochSeconds{ .secs = @bitCast(ts) };
    const day   = epoch.getEpochDay();
    const ds    = epoch.getDaySeconds();

    const yd   = day.calculateYearDay();
    const year = yd.year;
    const md   = yd.calculateMonthDay();
    const month = md.month.numeric();
    const day_n = md.day_index + 1;

    const hour   = ds.getHoursIntoDay();
    const minute = ds.getMinutesIntoHour();
    const second = ds.getSecondsIntoMinute();

    _ = std.fmt.bufPrint(
        out[0..20],
        "{d:0>4}-{d:0>2}-{d:0>2}T{d:0>2}:{d:0>2}:{d:0>2}Z",
        .{ year, month, day_n, hour, minute, second },
    ) catch {};
    out[20] = 0;
}

/// SHA-256 hex digest of `input` written into `out` (65 bytes: 64 hex + null).
pub fn sha256Hex(input: []const u8, out: *[65]u8) void {
    var digest: [32]u8 = undefined;
    std.crypto.hash.sha2.Sha256.hash(input, &digest, .{});
    for (digest, 0..) |b, i| {
        _ = std.fmt.bufPrint(out[i * 2 ..][0..2], "{x:0>2}", .{b}) catch {};
    }
    out[64] = 0;
}

/// Build a ProofEnvelope for the given response body and policy context.
/// Phase 1: light mode (SHA-256 hashes only). Signature field unused.
pub fn wrapResponse(body: []const u8, policy_context: []const u8) t.ProofEnvelope {
    var env: t.ProofEnvelope = std.mem.zeroes(t.ProofEnvelope);
    sha256Hex(body, &env.result_hash);
    sha256Hex(policy_context, &env.policy_hash);
    generateUuidV4(&env.query_id);
    formatRfc3339(&env.issued_at);
    @memcpy(env.proof_type[0..5], "light");
    env.proof_type[5] = 0;
    return env;
}

/// Serialise a ProofEnvelope to JSON, appending to `buf`.
/// Returns a slice into `buf` of the written bytes.
pub fn envelopeToJson(env: t.ProofEnvelope, buf: []u8) ![]const u8 {
    return try std.fmt.bufPrint(buf,
        "{{\"result_hash\":\"{s}\",\"policy_hash\":\"{s}\"," ++
        "\"query_id\":\"{s}\",\"issued_at\":\"{s}\"," ++
        "\"proof_type\":\"{s}\",\"signature\":\"\"}}",
        .{
            std.mem.sliceTo(&env.result_hash, 0),
            std.mem.sliceTo(&env.policy_hash, 0),
            std.mem.sliceTo(&env.query_id,    0),
            std.mem.sliceTo(&env.issued_at,   0),
            std.mem.sliceTo(&env.proof_type,  0),
        },
    );
}

/// Wrap a JSON body in a proof envelope, returning
///   {"data":<body>,"proof":<envelope>}
/// written into `out`. Returns a slice of the written bytes.
pub fn wrapBodyWithProof(
    body:           []const u8,
    policy_context: []const u8,
    out:            []u8,
) ![]const u8 {
    const env = wrapResponse(body, policy_context);

    var proof_buf: [512]u8 = undefined;
    const proof_json = try envelopeToJson(env, &proof_buf);

    return try std.fmt.bufPrint(out, "{{\"data\":{s},\"proof\":{s}}}", .{ body, proof_json });
}

/// Return the canonical policy context string for a module name.
/// Mirrors policy_context_string() from policy.v.
pub fn policyContextString(module_name: []const u8, out: []u8) ![]const u8 {
    return try std.fmt.bufPrint(
        out,
        "aerie-policy-v1:phase1-permissive:module={s}:entitlements=all",
        .{module_name},
    );
}

File: src/api/rust/src/proof.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// proof.rs — ProofEnvelope generation for Aerie API responses
//
// Mirrors src/api/v/proof.v (90 LOC).
//
// Every Aerie response is wrapped in a ProofEnvelope that commits the
// gateway to the exact bytes it served:
//
//   {
//     "data": <original response value>,
//     "proof": {
//       "result_hash":  "<sha256 of JSON-serialised data>",
//       "policy_hash":  "<sha256 of policy decision JSON>",
//       "query_id":     "<UUID v4 from PolicyDecision>",
//       "issued_at":    "<RFC 3339 timestamp>",
//       "proof_type":   "light",
//       "signature":    ""          ← Phase 2: Ed25519 over result_hash
//     }
//   }
//
// Phase 1 ("light" mode): hashes only, no cryptographic signature.
// The empty `signature` field is intentional — it is a placeholder for
// Phase 2 where the gateway signs responses with an Ed25519 key stored
// in a Stapeln secret.

use chrono::Utc;
use sha2::{Digest, Sha256};

/// The inner proof metadata block nested under the `"proof"` key.
#[derive(Debug, Clone, serde::Serialize)]
pub struct ProofMeta {
    /// SHA-256 hex digest of the JSON-serialised response data
    pub result_hash: String,
    /// SHA-256 hex digest of the policy decision JSON
    pub policy_hash: String,
    /// UUID v4 from the PolicyDecision — correlates request ↔ audit log ↔ proof
    pub query_id: String,
    /// RFC 3339 timestamp at envelope creation time
    pub issued_at: String,
    /// Always "light" in Phase 1; "full" when Ed25519 signature is added
    pub proof_type: String,
    /// Ed25519 signature over result_hash (Phase 2). Empty string in Phase 1.
    pub signature: String,
}

/// Top-level envelope that wraps every Aerie API response.
///
/// Callers receive `data` (the actual payload) alongside the proof
/// metadata, enabling independent verification of what the gateway
/// served and under what policy.
#[derive(Debug, Clone, serde::Serialize)]
pub struct ProofEnvelope {
    /// The original response payload — any JSON value
    pub data: serde_json::Value,
    /// Cryptographic commitment metadata
    pub proof: ProofMeta,
}

/// Wrap `data` in a ProofEnvelope.
///
/// `data` — the JSON value to wrap (already-serialised response body)
/// `policy_json` — the audit JSON from `decision_to_audit_json`
/// `query_id` — UUID v4 from the PolicyDecision for cross-correlation
///
/// Returns the fully-serialised envelope as a JSON string ready for
/// writing to the HTTP response body.
pub fn wrap_with_proof(
    data: serde_json::Value,
    policy_json: &str,
    query_id: &str,
) -> String {
    let result_hash = sha256_hex(&data.to_string());
    let policy_hash = sha256_hex(policy_json);
    let issued_at = Utc::now().to_rfc3339();

    let envelope = ProofEnvelope {
        data,
        proof: ProofMeta {
            result_hash,
            policy_hash,
            query_id: query_id.to_string(),
            issued_at,
            proof_type: "light".to_string(),
            signature: String::new(),
        },
    };

    serde_json::to_string(&envelope)
        .unwrap_or_else(|_| r#"{"error":"envelope_serialise_failed"}"#.to_string())
}

/// Compute the SHA-256 hex digest of a string.
fn sha256_hex(input: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(input.as_bytes());
    hex::encode(hasher.finalize())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn envelope_has_all_fields() {
        let data = serde_json::json!({"status": "ok"});
        let json = wrap_with_proof(data, r#"{"allowed":true}"#, "test-query-id");
        let v: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert!(v["data"].is_object());
        assert!(v["proof"]["result_hash"].is_string());
        assert!(v["proof"]["policy_hash"].is_string());
        assert_eq!(v["proof"]["query_id"], "test-query-id");
        assert_eq!(v["proof"]["proof_type"], "light");
        assert_eq!(v["proof"]["signature"], "");
    }

    #[test]
    fn result_hash_is_sha256_of_data() {
        let data = serde_json::json!({"x": 1});
        let json = wrap_with_proof(data.clone(), "{}", "qid");
        let v: serde_json::Value = serde_json::from_str(&json).unwrap();
        let expected = sha256_hex(&data.to_string());
        assert_eq!(v["proof"]["result_hash"], expected);
    }

    #[test]
    fn deterministic_hash_for_same_input() {
        let h1 = sha256_hex("hello aerie");
        let h2 = sha256_hex("hello aerie");
        assert_eq!(h1, h2);
        assert_eq!(h1.len(), 64); // 32 bytes → 64 hex chars
    }
}
